File: Assets/Scripts/new/Building/LaserCannon.cs
using UnityEngine;
using System.Collections;

[RequireComponent (typeof(CharacterStats))]
[RequireComponent (typeof(SphereCollider))]
public class LaserCannon : MonoBehaviour {

  public AudioClip laserSound;

  public GameObject laser;
  public float reloadTime;
  public float turningSpeed;

  private float damage;
  private float bonusDamage;
  private float attackingRange;

  public Transform[] muzzles;
  public Transform turretBall;

  private float nextFireTime;
  private float nextTargetTime;

  private Transform target;

  void Start() {
    CharacterStats characterStats = GetComponent<CharacterStats>();
    GetComponent<SphereCollider>().radius = characterStats.AttackingRange;

    damage = characterStats.Damage;
    bonusDamage = 0;

    target = null;

    nextTargetTime = nextFireTime = Time.time;
  }

  void Update() {
    if (target != null) {
      Quaternion desiredRotation = Quaternion.LookRotation(target.position - turretBall.position);
      turretBall.rotation = Quaternion.Slerp(turretBall.rotation, desiredRotation, Time.deltaTime * turningSpeed);
      if (Time.time >= nextFireTime) {
        FireProjectile();
      }
    }
  }

  void OnTriggerEnter(Collider collider) {
    if (target == null && collider.gameObject.tag == "Enemy") {
      target = collider.gameObject.transform;
    }
  }

  void OnTriggerStay(Collider collider) {
    if (target == null && collider.gameObject.tag == "Enemy") {
      target = collider.gameObject.transform;
    }
  }

  void OnTriggerExit(Collider collider) {
    if (target == collider.transform) {
      target = null;
    }
  }

  void FireProjectile() {
    AudioManager.PlayAudioClip(laserSound);

    nextFireTime = Time.time + reloadTime;

    for (int i = 0; i < muzzles.Length; ++i) {
      GameObject laserGameObject = Instantiate(laser, muzzles[i].position, Quaternion.identity) as GameObject;
      laserGameObject.GetComponent<Laser>().TargetPosition = target.position;
      CharacterStats targetCharacterStats = target.GetComponent<CharacterStats>();
      targetCharacterStats.CurrentHP -= (damage + bonusDamage);
      if (targetCharacterStats.CurrentHP <= 0) {
        /*
        if (Camera.main.GetComponent<MainFunction>().learnable) {
          GetComponent<CharacterStats>().kill++;
          bonusDamage = (GetComponent<Stats>().kill / 10.0f) / 100.0f * damage;
        }
        */
      }
    }
  }
}

File: Assets/old/Scripts/Buildings/LaserCannon.js
#pragma strict

private var damage : float;
private var bonusDamage : float;

var myLaser : GameObject;
var reloadTime : float;
var turnSpeed : float;
var firePauseTime : float;
var target : Transform;
var muzzlePositions : Transform[];
var turretBall : Transform;

private var nextFireTime : float;
private var nextMoveTime : float;
private var desiredRotation : Quaternion;

function Start () {
  damage = GetComponent(Stats).damage;
}

function Update () {
  if (target) {
    if (Time.time >= nextMoveTime) {
      CalculateAimPosition(target.position);
      turretBall.rotation = Quaternion.Lerp(turretBall.rotation, desiredRotation, Time.deltaTime * turnSpeed);
    }
    if (Time.time >= nextFireTime) {
      FireProjectile();
    }
  }
}

function OnTriggerEnter (other : Collider) {
  if (!target && other.gameObject.tag == "Enemy") {
    nextFireTime = Time.time + reloadTime * 0.5;
    target = other.gameObject.transform;
  }
}

function OnTriggerStay (other : Collider) {
  if (!target && other.gameObject.tag == "Enemy") {
    nextFireTime = Time.time + reloadTime * 0.5;
    target = other.gameObject.transform;
  }
}

function OnTriggerExit (other : Collider) {
  if (other.gameObject.transform == target) target = null;
}

function CalculateAimPosition (targetPos : Vector3) {
  var aimPoint = Vector3(targetPos.x, targetPos.y, targetPos.z);
  desiredRotation = Quaternion.LookRotation(aimPoint - transform.position) * Quaternion.Euler(0, -90, 0);
}

function FireProjectile () {
  GetComponent.<AudioSource>().Play();
  nextFireTime = Time.time + reloadTime;
  nextMoveTime = Time.time + firePauseTime;

  for (muzzle in muzzlePositions) {
    var laser : GameObject = Instantiate(myLaser, muzzle.position, muzzle.rotation);
    laser.GetComponent(Laser).targetPosition = target.position;
    target.GetComponent(Stats).nowHP -= (damage + bonusDamage);
    if (target.GetComponent(Stats).nowHP <= 0) {
      if (Camera.main.GetComponent(MainFunction).learnable) {
        GetComponent(Stats).kill++;
        bonusDamage = (GetComponent(Stats).kill / 10.0) / 100.0 * damage;
      }
    }
  }
}
